fsspec/implementations/cache_metadata.py

```diff
@@ -0,0 +1,210 @@
+                    )
+                fn = os.path.join(self._storage[-1], fn)
+                yield fn
+                self.cached_files[-1].pop(path)
+
+        if self.cached_files[-1]:
+            cache_path = os.path.join(self._storage[-1], "cache")
+            with atomic_write(cache_path) as fc:
+                pickle.dump(self.cached_files[-1], fc)
+
+    def close_file(self, f: Any, path: str) -> None:
+        """Perform side-effect actions when closing a cached file.
+
+        The actual closing of the file is the responsibility of the caller.
+        """
+        # File must be writeble, so in self.cached_files[-1]
+        c = self.cached_files[-1][path]
+        if c["blocks"] is not True and len(c["blocks"]) * f.blocksize >= f.size:
+            c["blocks"] = True
+
+    def empty(self) -> bool:
+        """Return ``True`` if metadata of the writable storage is empty"""
+        return not self.cached_files[-1]
+
+    def load(self) -> None:
+        """Load all metadata from disk and store in ``self.cached_files``"""
+        cached_files = []
+        for fn, _, _ in self._scan_locations():
+            if os.path.exists(fn):
+                with open(fn, "rb") as f:
+                    # TODO: consolidate blocks here
+                    loaded_cached_files = pickle.load(f)
+                    for c in loaded_cached_files.values():
+                        if isinstance(c["blocks"], list):
+                            c["blocks"] = set(c["blocks"])
+                    cached_files.append(loaded_cached_files)
+            else:
+                cached_files.append({})
+        self.cached_files = cached_files or [{}]
+
+    def pop_file(self, path: str) -> None:
+        details = self.check_file(path, None)
+        if not details:
+            return
+        _, fn = details
+        if fn.startswith(self._storage[-1]):
+            os.remove(fn)
+            self.cached_files[-1].pop(path)
+            self.save()
+        else:
+            raise PermissionError(
+                "Can only delete cached file in last, writable cache location"
+            )
+
+    def save(self) -> None:
+        """Save metadata to disk"""
+        for (fn, _, writable), cache in zip(self._scan_locations(), self.cached_files):
+            if not writable:
+                continue
+
+            if os.path.exists(fn):
+                with open(fn, "rb") as f:
+                    cached_files = pickle.load(f)
+                for k, c in cached_files.items():
+                    if k in cache:
+                        if c["blocks"] is True or cache[k]["blocks"] is True:
+                            c["blocks"] = True
+                        else:
+                            # self.cached_files[*][*]["blocks"] must continue to
+                            # point to the same set object so that updates
+                            # performed by MMapCache are propagated back to
```

Comment:
Shouldn't they already be the same object?

Reply:
I have moved this comment and surrounding code verbatim from current `master`: https://github.com/fsspec/filesystem_spec/blob/45a6aec7da1407243f9767c6ab0cff40efee72eb/fsspec/implementations/cached.py#L169-L172

I haven't yet invested any time into the block caching to have an opinion about this yet.